src/logics/morphoblue/logic.flash-loan.ts

```diff
@@ -0,0 +1,124 @@
+import { MorphoFlashLoanCallback__factory, Morpho__factory } from './contracts';
+import { Service } from './service';
+import * as common from '@protocolink/common';
+import * as core from '@protocolink/core';
+import { getContractAddress, getMarkets, supportedChainIds } from './configs';
+import invariant from 'tiny-invariant';
+
+export type FlashLoanLogicTokenList = common.Token[];
+
+export type FlashLoanLogicParams = core.FlashLoanParams;
+
+export type FlashLoanLogicQuotation = core.FlashLoanQuotation;
+
+export type FlashLoanLogicFields = core.FlashLoanFields;
+
+export class FlashLoanLogic extends core.Logic implements core.LogicTokenListInterface, core.LogicBuilderInterface {
+  static id = 'flash-loan';
+  static protocolId = 'morphoblue';
+  static readonly supportedChainIds = supportedChainIds;
+
+  get callbackAddress() {
+    return getContractAddress(this.chainId, 'MorphoFlashLoanCallback');
+  }
+
+  async calcCallbackFee(loan: common.TokenAmount) {
+    const callback = MorphoFlashLoanCallback__factory.connect(this.callbackAddress, this.provider);
+    const feeRate = await callback.feeRate();
+    const callbackFee = new common.TokenAmount(loan.token).setWei(common.calcFee(loan.amountWei, feeRate.toNumber()));
+
+    return callbackFee;
+  }
+
+  async getTokenList() {
+    const tokenList: FlashLoanLogicTokenList = [];
+    const service = new Service(this.chainId, this.provider);
+
+    const markets = getMarkets(this.chainId);
+
+    for (const market of markets) {
+      const loanTokens = await service.getLoanTokens(market.id);
+      tokenList.push(...loanTokens!);
+    }
+
+    return tokenList;
+  }
+
+  async quote(params: FlashLoanLogicParams) {
+    const assets = core.isFlashLoanLoanParams(params)
+      ? params.loans.map(({ token }) => token)
+      : params.repays.map(({ token }) => token);
+    invariant(assets.length === 1, 'flashLoan more than one token');
+
+    const morphoAddress = getContractAddress(this.chainId, 'Morpho');
+    const asset = assets[0];
+    const calls: common.Multicall3.CallStruct[] = [
+      {
+        target: asset.address,
+        callData: this.erc20Iface.encodeFunctionData('balanceOf', [morphoAddress]),
+      },
+    ];
+
+    const { returnData } = await this.multicall3.callStatic.aggregate(calls);
+
+    let j = 0;
+    const feeBps = 0;
+
+    let loans: common.TokenAmounts;
+    let repays: common.TokenAmounts;
+    if (core.isFlashLoanLoanParams(params)) {
+      ({ loans } = params);
+
+      repays = new common.TokenAmounts();
+      for (let i = 0; i < loans.length; i++) {
```

Comment:
`loans.length` should also be 1 since that `assets.length === 1`. If so, the `let j = 0` in L64 can also be simplified.